examples/tic-tac-toe/cli/src/client.rs

```diff
@@ -0,0 +1,749 @@
+    object::Owner,
+    programmable_transaction_builder::ProgrammableTransactionBuilder,
+    signature::GenericSignature,
+    transaction::{
+        InputObjectKind, ObjectArg, ProgrammableTransaction, Transaction, TransactionData,
+        TransactionKind,
+    },
+    Identifier,
+};
+
+use crate::{
+    crypto::combine_keys,
+    game::{self, Game, GameKind, Winner},
+    turn_cap::TurnCap,
+};
+
+#[derive(Parser, Debug)]
+pub struct Connection {
+    /// The Sui CLI config directory, (default: ~/.sui/sui_config/client.yaml)
+    #[clap(long)]
+    config: Option<PathBuf>,
+
+    /// Object ID of the game's package.
+    #[clap(long, short, env = "PKG")]
+    package_id: ObjectID,
+}
+
+pub(crate) struct Client {
+    wallet: WalletContext,
+    package: ObjectID,
+}
+
+impl Client {
+    /// Create a new client that derives its active address and RPC from the CLI's config (found at
+    /// path `config`), and that expects to interact with the tic-tac-toe package at address
+    /// `package`.
+    pub(crate) fn new(conn: Connection) -> Result<Self> {
+        let Some(config) = conn.config.or_else(|| {
+            let mut default = dirs::home_dir()?;
+            default.extend([".sui", "sui_config", "client.yaml"]);
+            Some(default)
+        }) else {
+            bail!("Can't find Wallet config");
+        };
+
+        let wallet = WalletContext::new(&config, None, None)?;
+        Ok(Self {
+            wallet,
+            package: conn.package_id,
+        })
+    }
+
+    /// Fetch the details of a game object from on-chain (can be either shared or owned).
+    pub(crate) async fn game(&self, id: ObjectID) -> Result<Game> {
+        let client = self.client().await?;
+
+        // (1) Read from RPC
+        let response = client
+            .read_api()
+            .get_object_with_options(
+                id,
+                SuiObjectDataOptions {
+                    show_owner: true,
+                    show_bcs: true,
+                    ..Default::default()
+                },
+            )
+            .await
+            .context("Error fetching game over RPC")?;
+
+        if let Some(err) = response.error {
+            bail!(err);
+        }
+
+        // (2) Perform validation checks
+        let Some(SuiObjectData {
+            object_id,
+            version,
+            digest,
+            bcs: Some(raw),
+            owner: Some(owner),
+            ..
+        }) = response.data
+        else {
+            bail!("No data for game");
+        };
+
+        let Some(raw) = raw.try_as_move() else {
+            bail!("It is a package, not a object.");
+        };
+
+        if raw.type_.name.as_str() != "Game" {
+            bail!("It is not a Game object, it has type {}", raw.type_);
+        }
+
+        let package = ObjectID::from(raw.type_.address);
+        if package != self.package {
+            bail!(
+                "It is expected to be from package {} but has type {}",
```

Comment:
It can be one of two types `$PKG::owned::Game` or `$PKG::shared::Game` -- maybe the most straightforward thing is to compare apples to apples, i.e. to say:

```suggestion
                "It is expected to be from package {} but is from package {}",
```